crates/state/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+                    FROM "values"
+                    WHERE "key_hash" = ?1
+                    ORDER BY "id" DESC
+                    LIMIT 1
+                "#,
+                hash,
+            )
+            .fetch_optional(&self.sqlite)
+            .await?
+            .map(|record| record.value);
+            Ok(value)
+        }
+    }
+
+    /// Sets the `value` associated with the Blake3 hash of `key`. See
+    /// [`GlobalState::upsert_by_hash`] for more information.
+    pub async fn upsert<V>(&mut self, key: &[u8], value: V)
+    where
+        V: Into<Vec<u8>>,
+    {
+        let hash = Blake3Hasher::hash(key);
+        self.upsert_by_hash(hash, value).await;
+    }
+
+    /// Sets the `value` associated with a Blake3 `hash`. This change will be "dirty" until
+    pub async fn upsert_by_hash<V>(&mut self, hash: [u8; 32], value: V)
+    where
+        V: Into<Vec<u8>>,
+    {
+        let value = value.into();
+        self.dirty_changes.entry(hash).or_insert(value);
+    }
+
+    pub async fn merkelize(&self) -> Option<Commit> {
+        None
+    }
+
+    /// Creates a new [`Commit`] with the given signature.
+    async fn create_commit(&mut self, signature: Commit) -> Result<()> {
+        let signature = &signature[..];
+        sqlx::query!(
+            r#"
+                INSERT INTO "commits" ("id", "signature")
+                VALUES (?1, ?2)
+            "#,
+            self.next_commit_id,
+            signature
+        )
+        .execute(&self.sqlite)
+        .await?;
+        Ok(())
+    }
+
+    pub async fn commit(&mut self) -> Result<Commit> {
+        let commit = self.checkpoint().await?;
+        self.next_commit_id += 1;
+        self.create_commit(commit).await?;
+        Ok(commit)
+    }
+
+    /// Persists all dirty changes from memory to disk. Even though persisted to
+    /// disk, they still don't belong to any commit; look into
+    /// [`GlobalState::commit`].
+    pub async fn checkpoint(&mut self) -> Result<Commit> {
+        let dirty_changes = std::mem::take(&mut self.dirty_changes);
+
+        let mut signature = [0; 32];
+
+        for change in dirty_changes {
+            let old_value: Option<Vec<u8>> = self.get_by_hash(&change.0, None).await?;
+            if let Some(old_value) = old_value {
+                xor_signature(&mut signature, &hash_key_value_pair(&change.0, &old_value));
+            }
+            xor_signature(&mut signature, &hash_key_value_pair(&change.0, &change.1));
+
+            let key_hash = &change.0[..];
+            let value = &change.1[..];
+            sqlx::query!(
+                r#"
+                    INSERT INTO "values" ("commit_id", "key_hash", "value")
+                    VALUES (?1, ?2, ?3)
+                "#,
+                self.next_commit_id,
+                key_hash,
+                value
+            )
+            .execute(&self.sqlite)
+            .await?;
+        }
+
+        self.update_commit_signature(&signature).await
```

Comment:
If we are only inserting new `(key, value, next_commit_id) why do we want to call the `update_commit_signature`
?

Reply:
Any specific reason why we don't want to write to SQLite on `checkpoint`? Saving state on `checkout` is consistent with what `git` and other systems generally mean by that word. What does a `Journal` (or `dirty_tx_changes` / `dirty_commit_changes`) achieve?

Reply:
Solved, we're not writing anything to db when using `checkpoint`.